.scripts/get-interface-test-targets.py

```diff
@@ -0,0 +1,145 @@
+"""
+
+import argparse
+import json
+import logging
+import pathlib
+import re
+import subprocess
+import tempfile
+
+import yaml
+
+_REPO_ROOT = pathlib.Path(__file__).parent.parent
+
+logging.basicConfig(level=logging.DEBUG)
+logger = logging.getLogger(str(pathlib.Path(__file__).relative_to(_REPO_ROOT)))
+
+
+def _main() -> None:
+    parser = argparse.ArgumentParser()
+    parser.add_argument('interface', help='Path from repo root to specific interface directory.')
+    parser.add_argument('--all', action='store_true', help='Include combinations with no tests.')
+    args = parser.parse_args()
+    targets = _target_from_interface(args.interface, has_tests_only=not args.all)
+    output = json.dumps(targets)
+    logger.info(output)
+    print(output)
+
+
+def _target_from_interface(interface_str: str, has_tests_only: bool) -> list[dict[str, str]]:
+    """Return a list of interface test targets for this interface.
+
+    Iterates over the interface versions, collects all provider and requirer charms, and clones
+    the charms to discover their endpoints.
+
+    Args:
+        interface_str: The interface path relative to the repo root, e.g. 'interfaces/tracing'.
+        has_tests_only: If true, exclude any targets that don't actually have any tests defined.
+
+    Returns:
+        A list of interface test targets. Each target is a dictionary containing:
+            - interface (name)
+            - version (with v prefix)
+            - role (provide or require)
+            - charm (name)
+            - endpoint (name)
+    """
+    interface = _REPO_ROOT / interface_str
+    targets: list[dict[str, str]] = []
+    for v in sorted((interface / 'interface').glob('v[0-9]*')):
+        interface_yaml = yaml.safe_load((v / 'interface.yaml').read_text())
+        for role in 'provide', 'require':
+            charms = interface_yaml.get(f'{role}rs', [])
+            if not charms:
+                logger.debug('No charms for %s %s %s role.', interface_str, v.name, role)
+                continue
+            if has_tests_only and not _has_tests(v, f'{role}r'):
+                msg = 'Skipping these charms because there are no tests for %s %s %s: %s'
+                logger.warning(msg, interface_str, v.name, role, charms)
+                continue
+            targets.extend(
+                {
+                    'interface': interface.name,
+                    'version': v.name,
+                    'role': role,
+                    'charm_name': charm['name'],
+                    'endpoint': endpoint,
+                }
+                for charm in charms
+                for endpoint in _get_endpoints(
+                    interface=interface.name,
+                    role=f'{role}s',
+                    charm_repo=charm['url'],
+                    charm_ref=charm.get('branch', 'main'),
+                    charm_root=charm.get('test_setup', {}).get('charm_root', ''),
+                )
+            )
```

Comment:
Let's avoid the nested comprehension (see our [style guide](https://github.com/canonical/operator/blob/main/STYLE.md#avoid-nested-comprehensions-and-generator-expressions)). I also prefer avoiding line breaks in the `in` expression of a for loop (like how `_get_endpoints` is broken), for example:

```python
for charm in charms:
    endpoint = _get_endpoints(
        ...
    )
    for endpoint in endpoints:
        target = {
            ...
        }
        targets.append(target)
```